wandb/sdk/internal/sender.py

```diff
@@ -991,12 +997,56 @@ def finish(self):
             self._fs = None
 
     def _max_cli_version(self):
-        _, server_info = self._api.viewer_server_info()
+        server_info = self.get_server_info()
         max_cli_version = server_info.get("cliVersionInfo", {}).get(
             "max_cli_version", None
         )
         return max_cli_version
 
+    def get_viewer_server_info(self):
+        if not self._cached_server_info or not self._cached_viewer:
+            (
+                self._cached_viewer,
+                self._cached_server_info,
+            ) = self._api.viewer_server_info()
+
+    def get_viewer_info(self):
+        if not self._cached_viewer:
+            self.get_viewer_server_info()
+        return self._cached_viewer
+
+    def get_server_info(self):
+        if not self._cached_server_info:
+            self.get_viewer_server_info()
+        return self._cached_server_info
+
+    def get_local_info(self):
+        """
+        This is a helper function that queries the server to get the the local version information.
+        First, we perform an introspection, if it returns empty we deduce that the docker image is
+        out-of-date. Otherwise, we use the returned values to deduce the state of the local server.
+        """
+        local_info = wandb_internal_pb2.LocalInfo()
+
+        latest_local_version = "latest"
+
+        if len(self._api.server_info_introspection()) == 0:
```

Comment:
Is this correct?

`server_info_introspection` can return three values:

1. Empty dict `{}`
2. negative result (when `LocalVersionInfo` doesn't exist):
```
{
  __type: null
}
```

3. positive (and potentially more variations depending on history of this type)
```
{
  _type: {
    fields: [ { name: 'x' }, { name: 'y' } ], ... ]
  }
}
```

This check catches the first possible r/v, but misses the 2nd.

Reply:
wait, i have an else case